File: java/ClojurePH.java
package irresponsible.crouton;

import clojure.lang.IFn;
import clojure.lang.IPersistentVector;
import clojure.lang.ITransientMap;

public class ClojurePH extends PH {
  private final IFn ifn;
  public ClojurePH(Object n, IFn i, IRoute next) {
    super(n,next);
    if (null == i)
      throw new IllegalArgumentException("lambda must not be nil");
    ifn = i;
  }
  public final Object match(IPersistentVector pieces, ITransientMap places) {
    String piece = (String) pieces.nth(0, null);
    if (piece == null) return null;
    Object ret = ifn.invoke(piece);
    return (ret != null) ? forward(piece, pieces, places) : null;
  }
}

File: kotlin/src/ClojurePH.kt
package irresponsible.crouton

import clojure.lang.IFn
import clojure.lang.IPersistentVector
import clojure.lang.IPersistentMap

class ClojurePH(n: Any, private val ifn: IFn, next: IRoute) : PH(n, next) {
    override fun match(pieces: IPersistentVector, places: IPersistentMap): Any? {
      val piece = pieces.nth(0, null)
      if (piece == null) return null
      val piece2 = piece as String
      val ret = ifn.invoke(piece)
      return if (ret != null) forward(piece, pieces, places) else null
    }
}
